python_modules/dagster/dagster/_core/external_execution/context.py

```diff
@@ -0,0 +1,80 @@
+from typing import Optional
+
+from dagster_external.protocol import (
+    ExternalDataProvenance,
+    ExternalExecutionContextData,
+    ExternalExecutionUserdata,
+    ExternalTimeWindow,
+)
+
+from dagster._core.definitions.data_version import DataProvenance
+from dagster._core.definitions.events import AssetKey
+from dagster._core.definitions.partition_key_range import PartitionKeyRange
+from dagster._core.definitions.time_window_partitions import TimeWindow
+from dagster._core.execution.context.compute import OpExecutionContext
+
+
+def build_external_execution_context(
+    context: OpExecutionContext,
+    userdata: Optional[ExternalExecutionUserdata],
+) -> "ExternalExecutionContextData":
+    asset_key = context.asset_key_for_output()
+    asset_provenance = context.get_asset_provenance(asset_key)
+    partition_key = context.partition_key if context.has_partition_key else None
+    partition_time_window = context.partition_time_window if context.has_partition_key else None
+    partition_key_range = context.partition_key_range if context.has_partition_key else None
+    return ExternalExecutionContextData(
+        asset_key=_convert_asset_key(asset_key),
+        code_version=context.job_def.asset_layer.code_version_for_asset(asset_key),
+        data_provenance=(_convert_data_provenance(asset_provenance) if asset_provenance else None),
+        partition_key=partition_key,
+        partition_key_range=(
+            _convert_partition_key_range(partition_key_range) if partition_key_range else None
+        ),
+        partition_time_window=(
+            _convert_time_window(partition_time_window) if partition_time_window else None
+        ),
+        run_id=context.run_id,
+        run_tags=context.run_tags,
```

Comment:
on further thought - I think tags is sort of like config in that we can leave it to the user to pass through what they care about as userdata and omit it from the default set

Reply:
ok I'll address this in the followup